bot/exts/fun/hangman.py

```diff
@@ -0,0 +1,144 @@
+    async def hangman(
+            self,
+            ctx: commands.Context,
+            min_length: str = "0",
+            max_length: str = "25",
+            min_unique_letters: str = "0",
+            max_unique_letters: str = "25",
+            singleplayer: Literal["s", "m"] = "s",
+    ) -> None:
+        """
+        Play hangman against the bot, where you have to guess the word it has provided!
+
+        The arguments for this command mean:
+        - min_length: the minimum length you want the word to be (i.e. 2)
+        - max_length: the maximum length you want the word to be (i.e. 5)
+        - min_unique_letters: the minimum unique letters you want the word to have (i.e. 4)
+        - max_unique_letters: the maximum unique letters you want the word to have (i.e. 7)
+        - singleplayer: writing 's' means you want to play by yourself, and only you can suggest letters,
+            - writing 'm' means you want multiple players to join in and guess the word.
+        """
+        # filtering the list of all words depending on the configuration
+        filtered_words = [
+            word for word in ALL_WORDS
+            if int(min_length) < len(word) < int(max_length)
+            and int(min_unique_letters) < len(set(word)) < int(max_unique_letters)
+        ]
+
+        if not filtered_words:
+            filter_not_found_embed = Embed(
+                title=choice(NEGATIVE_REPLIES),
+                description="No words could be found that fit all the filters specified.",
+                color=Colours.soft_red,
+            )
+            await ctx.send(embed=filter_not_found_embed)
+            return
+
+        # a dictionary mapping the images of the "hung man" to the number of tries it corresponds to
+        mapping_of_images = {tries_key: image_name for tries_key, image_name in zip(range(6, -1, -1), IMAGES)}
+        word = choice(filtered_words)
+
+        user_guess = "_" * len(word)
+        tries = 6
+
+        hangman_embed = Embed(
+            title="Hangman",
+            color=Colours.python_blue,
+        )
+        hangman_embed.set_image(url=mapping_of_images[tries])
+        hangman_embed.add_field(
+            name=f"The word is `{user_guess}`",
+            value="Guess the word by sending a message with the letter!",
+            inline=False,
+        )
+        original_message = await ctx.send(embed=hangman_embed)
+
+        while user_guess != word:
+            try:
+                message = await self.bot.wait_for(
+                    event="message",
+                    timeout=30.0,
+                    check=lambda msg: msg.author != self.bot if singleplayer == 'm' else msg.author == ctx.author
+                )
+            except TimeoutError:
+                return
+
+            if len(message.content) > 1:
+                continue
+
+            elif message.content in word:
+                positions = [idx for idx, letter in enumerate(word) if letter == message.content]
```

Comment:
This could be a set comprehension because we do a lot of `in` checks later on.
```suggestion
                positions = {idx for idx, letter in enumerate(word) if letter == message.content}
```